feat(coderd): add overload protection with rate limiting and concurrency control

kacpersaw opened the pull request:
## Summary

This adds configurable overload protection to the AI Bridge daemon to prevent the server from being overwhelmed during periods of high load.

Partially addresses coder/internal#1153 (rate limits and concurrency control; circuit breakers are deferred to a follow-up).

## New Configuration Options

| Option | Environment Variable | Description | Default |
|--------|---------------------|-------------|---------|
| `--aibridge-max-concurrency` | `CODER_AIBRIDGE_MAX_CONCURRENCY` | Maximum number of concurrent AI Bridge requests. Set to 0 to disable (unlimited). | `0` |
| `--aibridge-rate-limit` | `CODER_AIBRIDGE_RATE_LIMIT` | Maximum number of AI Bridge requests per second. Set to 0 to disable rate limiting. | `0` |

## Behavior

When limits are exceeded:
- **Concurrency limit**: Returns HTTP `503 Service Unavailable` with message "AI Bridge is currently at capacity. Please try again later."
- **Rate limit**: Returns HTTP `429 Too Many Requests` with `Retry-After` header.

Both protections are optional and disabled by default (0 values).

## Implementation

The overload protection is implemented as reusable middleware in `coderd/httpmw/ratelimit.go`:

1. **`RateLimitByAuthToken`**: Per-user rate limiting that uses `APITokenFromRequest` to extract the authentication token, with fallback to `X-Api-Key` header for AI provider compatibility (e.g., Anthropic). Falls back to IP-based rate limiting if no token is present. Includes `Retry-After` header for backpressure signaling.
2. **`ConcurrencyLimit`**: Uses an atomic counter to track in-flight requests and reject when at capacity.

The middleware is applied in `enterprise/coderd/aibridge.go` via `r.Group` in the following order:
1. Concurrency check (faster rejection for load shedding)
2. Rate limit check

**Note**: Rate limiting currently applies to all AI Bridge requests, including pass-through requests. Ideally only actual interceptions should count, but this would require changes in the aibridge library.

## Testing

Added comprehensive tests for:
- Rate limiting by auth token (Bearer token, X-Api-Key, no token fallback to IP)
- Different tokens not rate limited against each other
- Disabled when limit is zero
- Retry-After header is set on 429 responses
- Concurrency limiting (allows within limit, rejects over limit, disabled when zero)

kacpersaw committed:
feat(aibridged): add overload protection with rate limiting and concurrency control

This adds configurable overload protection to the AI Bridge daemon to prevent
the server from being overwhelmed during periods of high load.

New configuration options:
- CODER_AIBRIDGE_MAX_CONCURRENCY: Maximum number of concurrent AI Bridge requests (0 to disable)
- CODER_AIBRIDGE_RATE_LIMIT: Maximum number of requests per rate window (0 to disable)
- CODER_AIBRIDGE_RATE_WINDOW: Duration of the rate limiting window (default: 1m)

When limits are exceeded:
- Concurrency limit: Returns HTTP 503 Service Unavailable
- Rate limit: Returns HTTP 429 Too Many Requests

The overload protection is implemented as middleware in coderd/httpmw/ratelimit.go
and applied to the aibridged HTTP handler:
- RateLimitByIP: Rate limiting keyed by IP address for endpoints without API key context
- ConcurrencyLimit: Limits concurrent requests using an atomic counter

Both protections are optional and disabled by default (0 values).

Fixes coder/internal#1153

dannykopping reviewed (changes requested):
I'm not a fan of IP-based rate-limiting. It's brittle and lacks the specificity we need.

Limits need to apply per user. We could do that by using the API key in the request.

Also, a nit, your description says:

> Fixes coder/internal#1153

But it only addresses the first two points AFAICS.

kacpersaw committed:
refactor: use per-user rate limiting by auth token instead of IP

Address review feedback from @dannykopping:

1. Changed RateLimitByIP to RateLimitByAuthToken which extracts the
   authentication token from the Authorization header (Bearer token) or
   X-Api-Key header. Falls back to IP-based rate limiting if no token
   is present.

2. Added Retry-After header to 429 responses for better backpressure
   signaling.

3. Added comprehensive tests for the new auth token-based rate limiting
   including tests for Bearer token, X-Api-Key, different tokens, and
   fallback to IP.

kacpersaw committed:
refactor: address review feedback from @pawbana

Changes:
1. Remove --aibridge-rate-window config option, hardcode to 1 minute for simpler
   configuration (rate limit is now 'requests per minute').

2. Use APITokenFromRequest instead of custom extractAuthToken function, with
   fallback to X-Api-Key header for AI provider compatibility (Anthropic).

3. Simplify rate limit error message to 'You've been rate limited. Please try
   again later.' (Retry-After header provides timing info).

4. Add TODO comment noting that rate limiting currently applies to all AI Bridge
   requests including pass-through. Ideally only interceptions should count,
   but this requires changes in the aibridge library.

5. Consolidate rate limit tests with table-driven approach and use time.Hour
   instead of time.Second for more reliable test behavior.

kacpersaw committed:
refactor: address remaining review comments

1. Move rate limiters to r.Group middleware instead of manually wrapping
   in HandleFunc (aibridge.go).

2. Improve concurrency test synchronization by using sync.WaitGroup as a
   barrier to ensure all requests are in the handler before testing rejection.
   Also use close(releaseHandler) instead of sending to each goroutine.

dannykopping commented on enterprise/coderd/aibridge.go:
```diff
@@ -23,36 +24,55 @@ import (
+	// require changes in the aibridge library where the interception decision is made.
+	bridgeCfg := api.DeploymentValues.AI.BridgeConfig
+	concurrencyLimiter := httpmw.ConcurrencyLimit(bridgeCfg.MaxConcurrency.Value(), "AI Bridge")
+	rateLimiter := httpmw.RateLimitByAuthToken(int(bridgeCfg.RateLimit.Value()), aiBridgeRateLimitWindow)
```
I'm curious why we need an additional rate-limiter over and above what coderd already supports.

This results in confusing response headers:

```
< HTTP/1.1 503 Service Unavailable
< Content-Type: application/json; charset=utf-8
< Set-Cookie: csrf_token=NLIScrORyqJt0GRZw7dgvGmZL7G9uYoPnPKPgZscqGI=; Path=/; HttpOnly; SameSite=Lax
< Traceparent: 00-9f142cbfe9ed17484ab996f0fde3a13c-9d3425845043ec9d-01
< Vary: Origin
< Vary: Cookie
< X-Coder-Build-Version: v2.29.0-devel+72471fcfc
< X-Coder-Request-Id: 223e8a8b-021f-4d70-890a-ac24181c6b77
< X-Content-Type-Options: nosniff
< X-Ratelimit-Limit: 512                           <-----
< X-Ratelimit-Remaining: 509                 <-----
< X-Ratelimit-Reset: 1765274580
< X-Span-Id: 9d3425845043ec9d
< X-Trace-Id: 9f142cbfe9ed17484ab996f0fde3a13c
< Date: Tue, 09 Dec 2025 10:02:14 GMT
< Content-Length: 74
<
{"message":"AI Bridge is currently at capacity. Please try again later."}
```

Did you try to make aibridge work with the existing rate-limiter?

dannykopping commented on coderd/httpmw/ratelimit.go:
```diff
@@ -70,3 +72,70 @@ func RateLimit(count int, window time.Duration) func(http.Handler) http.Handler
+			// (e.g., Anthropic uses this header).
+			if apiKey := r.Header.Get("X-Api-Key"); apiKey != "" {
+				return strings.TrimSpace(apiKey), nil
+			}
```
This logic already exists in `enterprise/aibridged/http.go` -> `ExtractAuthToken`.

dannykopping commented on codersdk/deployment.go:
```diff
@@ -3391,6 +3391,26 @@ Write out the current server config as YAML to stdout.`,
+		},
+		{
+			Name:        "AI Bridge Rate Limit",
+			Description: "Maximum number of AI Bridge requests per minute. Set to 0 to disable rate limiting.",
```
Nit: consistency.

```suggestion
			Description: "Maximum number of AI Bridge requests per minute. Set to 0 to disable (unlimited).",
```

pawbana commented on cli/testdata/coder_server_--help.golden:
```diff
@@ -129,11 +129,8 @@ AI BRIDGE OPTIONS:
-      --aibridge-rate-window duration, $CODER_AIBRIDGE_RATE_WINDOW (default: 1m)
-          Duration of the rate limiting window for AI Bridge requests.
+          Maximum number of AI Bridge requests per minute. Set to 0 to disable
+          rate limiting.
```
nit: sorry for being pedantic and not thinking before but maybe limit per second would be easier to use then per minute?

kacpersaw committed:
refactor: change rate limit window from per-minute to per-second

Per @pawbana's feedback, per-second is easier to reason about than per-minute.

dannykopping commented on coderd/httpmw/ratelimit_test.go:
```diff
@@ -145,3 +146,202 @@ func TestRateLimit(t *testing.T) {
+		server := httptest.NewServer(rtr)
+		defer server.Close()
+
+		ctx := t.Context()
```
Needs a timeout

dannykopping reviewed (commented):
Overall LGTM. Tested it against my deployment and it works as expected.

Can you please add an integration test (one per mechanism) just so we can be assured that this works with the whole API stack?

dannykopping commented:
@kacpersaw I also think it's worth documenting how/why this differs from the normal API rate-limiting.

kacpersaw committed:
refactor: address @dannykopping's review comments

1. Add 'per replica' to description for both max concurrency and rate limit
   to clarify these limits are replica-local.

2. Use consistent '(unlimited)' wording for disabled state in rate limit
   description.

3. Remove TODO comment about pass-through requests - they should be rate
   limited since they still consume resources.

4. Add timeout to concurrency test using testutil.WaitShort.

kacpersaw commented:
@dannykopping Regarding your question about `>` vs `>=` on line 132:

Using `>` is correct for "max concurrent" semantics. Here's why:

```go
c := current.Add(1)  // Returns NEW value after incrementing
defer current.Add(-1)

if c > maxConcurrent {  // Reject if over limit
```

With `maxConcurrent = 2`:
- Request 1: `c = 1`, check `1 > 2` → false → **allowed** (1 concurrent)
- Request 2: `c = 2`, check `2 > 2` → false → **allowed** (2 concurrent)  
- Request 3: `c = 3`, check `3 > 2` → true → **rejected** ✓

If we used `>=`:
- Request 1: `c = 1`, check `1 >= 2` → false → allowed
- Request 2: `c = 2`, check `2 >= 2` → true → **rejected** ✗ (we want 2 concurrent allowed!)

The test `TestConcurrencyLimit/LimitsConcurrentRequests` validates this behavior - it sets `maxConcurrency = 2`, starts 2 requests that block in handlers, then verifies the 3rd request gets 503.

kacpersaw committed:
docs: document how RateLimitByAuthToken differs from RateLimit

Add detailed documentation explaining the key differences:
- Extracts token from headers vs context (for internal auth handling)
- No bypass header support for Owners
- No per-endpoint keying (limit applies across all endpoints)
- Includes Retry-After header for backpressure

kacpersaw commented:
Regarding the `ExtractAuthToken` duplication comment:

We can't import `enterprise/aibridged` from `coderd/httpmw` due to the AGPL licensing restriction (AGPL code cannot import enterprise packages, as enforced by `scripts/check_enterprise_imports.sh`).

The current approach uses `APITokenFromRequest` from `coderd/httpmw/apikey.go` (which handles Bearer tokens, cookies, and standard Coder auth methods) with a small fallback for `X-Api-Key` header for AI provider compatibility (Anthropic uses this header).

The X-Api-Key fallback is only 3 lines, so the duplication is minimal and unavoidable given the package boundaries.

kacpersaw committed:
refactor: remove ExtractAuthToken alias, use httpmw.ExtractAPIKeyFromHeader directly

kacpersaw committed:
fix lint

dannykopping commented on coderd/httpmw/apikey.go:
```diff
@@ -812,3 +812,24 @@ func CustomRedirectToLogin(rw http.ResponseWriter, r *http.Request, redirect str
+// This differs from [APITokenFromRequest] in that it only checks headers commonly
+// used by AI providers, not Coder-specific authentication methods like cookies
+// or the Coder-Session-Token header.
+func ExtractAPIKeyFromHeader(header http.Header) string {
```
If we move this here, we have to name this very carefully because it's only used in specific situations (i.e. aibridge).

I'd suggest moving this into a non-enterprise aibridge package.

dannykopping reviewed (approved):
LGTM once `ExtractAPIKeyFromHeader` is named/moved appropriately,

kacpersaw committed:
Use testutil.Context

kacpersaw committed:
refactor: move ExtractAuthToken to coderd/aibridge package

Creates a new non-enterprise coderd/aibridge package for AI Bridge
utilities. This properly encapsulates the auth token extraction function
that's specific to AI provider authentication patterns (Bearer token and
X-Api-Key headers).

- Add coderd/aibridge/aibridge.go with ExtractAuthToken function
- Update coderd/httpmw/ratelimit.go to use aibridge.ExtractAuthToken
- Update enterprise/aibridged to use aibridge.ExtractAuthToken
- Remove ExtractAPIKeyFromHeader from coderd/httpmw/apikey.go

kacpersaw committed:
test: add integration tests for AI Bridge rate and concurrency limiting

Adds two integration tests that verify the overload protection middleware
works correctly with the full API stack:

- TestAIBridgeRateLimiting: Verifies requests are rate limited (429) after
  exceeding the configured limit, with Retry-After header set.
- TestAIBridgeConcurrencyLimiting: Verifies concurrent requests beyond the
  limit are rejected (503) while allowing requests up to the limit.

kacpersaw committed:
fix: avoid require in goroutine to prevent data races

dannykopping commented on enterprise/aibridged/aibridged_test.go:
```diff
@@ -17,6 +17,7 @@ import (
 
 	"cdr.dev/slog/sloggers/slogtest"
 	"github.com/coder/aibridge"
+	coderaibridge "github.com/coder/coder/v2/coderd/aibridge"
```
Nit: the common pattern is to name this like `agplaibridge`.
See `enterprise/cli/root.go`.

kacpersaw committed:
refactor: update import alias for aibridge package

Commit:
Merge branch 'main' into kacpersaw/aibridge-overload-protection
